### DISCUSSION TITLE
Is it possible to include an exception to a glob exclusion?

### DISCUSSION THREAD

**Discussion by jpmor** [BurntSushi/ripgrep#1645] [General] [answered]
If I have a directory `foo/` with a bunch of subdirectories, one of which is `bar/`...  Is there a way to ensure `bar/` is the only directory searched _in foo/_ via globbing?

Maybe something like: `-g '!foo/!(bar/)'`

To make the example more concrete, I am working with a code base with many directories in `vendor/`, and I want to exclude all of them _except one or two_ from searches.

Thanks!

**Comment by BurntSushi** [comment, marked as answer]
@jpmor The simplest solution would be to just specify those directories to search, e.g., `-g 'foo/bar/**'`. But if that doesn't work, then you could do `-g 'foo/**' -g '!foo/*' -g 'foo/bar'` or something like that. But the simplest way to express "search many directories but exclude all except one or two" is to just specify the one or two things you want to search.

**Reply by jpmor** [reply]
@BurntSushi - thanks for the quick response!  Making sure things outside of `foo` are also searched was the complicating factor, but based on your suggestion I've tried:
```
-g '**' -g '!foo/*' -g 'foo/bar'
```
That's giving me exactly what I was looking for, thank you!

**Reply by jpmor** [reply]
Actually, here's an (uglier) but even better option:
```
rg -g '!{'"$(echo foo/!(bar) | sed 's/ /,/g')"'}/**'
```
This just shifts the matching onto bash `extglob`  or zsh `extended_glob` etc.

This is much faster that having an initial `-g '**'`.
